### PR TITLE
Handle TableMultiVariableLookup::points

### PR DESCRIPTION
**Pull request by jmarrec**
Related to discussion with @macumber  in #3223 

```ruby
m = OpenStudio::Model::Model.new
capModFuncOfWaterFlow = OpenStudio::Model::TableMultiVariableLookup.new(m, 1)
capModFuncOfWaterFlow.setName("CapModFuncOfWaterFlow")
capModFuncOfWaterFlow.setCurveType("Quadratic")
capModFuncOfWaterFlow.setInterpolationMethod("EvaluateCurveToLimits")
capModFuncOfWaterFlow.setMinimumValueofX1(0)
capModFuncOfWaterFlow.setMaximumValueofX1(1.33)
capModFuncOfWaterFlow.setMinimumTableOutput(0.0)
capModFuncOfWaterFlow.setMaximumTableOutput(1.04)
capModFuncOfWaterFlow.setInputUnitTypeforX1("Dimensionless")
capModFuncOfWaterFlow.setOutputUnitType("Dimensionless")

capModFuncOfWaterFlow.addPoint(0.0, 0.0)
capModFuncOfWaterFlow.addPoint(0.05, 0.001)
capModFuncOfWaterFlow.addPoint(0.33333, 0.71)
capModFuncOfWaterFlow.addPoint(0.5, 0.85)
capModFuncOfWaterFlow.addPoint(0.666667, 0.92)
capModFuncOfWaterFlow.addPoint(0.833333, 0.97)
capModFuncOfWaterFlow.addPoint(1.0, 1.0)
capModFuncOfWaterFlow.addPoint(1.333333, 1.04)

points = capModFuncOfWaterFlow.points
```

**Original behavior**:

```ruby
points
=> #<SWIG::TYPE_p_std__vectorT_std__pairT_std__vectorT_double_std__allocatorT_double_t_t_double_t_std__allocatorT_std__pairT_std__vectorT_double_std__allocatorT_double_t_t_double_t_t_t:0x007f9688250ac8
 @__swigtype__=
  "_p_std__vectorT_std__pairT_std__vectorT_double_std__allocatorT_double_t_t_double_t_std__allocatorT_std__pairT_std__vectorT_double_std__allocatorT_double_t_t_double_t_t_t">

points[0]
NoMethodError: undefined method `[]' for #<SWIG::TYPE_p_std__vectorT_std__pairT_std__vectorT_double_std__allocatorT_double_t_t_double_t_std__allocatorT_std__pairT_std__vectorT_double_std__allocatorT_double_t_t_double_t_t_t:0x007f9688250ac8>
from (pry):41:in `__pry__'
```

**With the changes**

```ruby
[23] OS-build(main)> points
=> [[[0.0], 0.0],
 [[0.05], 0.001],
 [[0.33333], 0.71],
 [[0.5], 0.85],
 [[0.666667], 0.92],
 [[0.833333], 0.97],
 [[1.0], 1.0],
 [[1.333333], 1.04]]
[24] OS-build(main)> points[0]
=> [[0.0], 0.0]
```

### CODE REVIEW

**Comment by macumber** [comment]
Waiting on reply by @jmarrec

**Comment by jmarrec** [comment]
I’ll do that if you insist, but just like we template a std::vector<double> as “DoubleVector” we could just use generic names for these templates.

——
Sent from a mobile device, please excuse the brevity.
Julien Marrec, EBCP, BPI MFBA
Owner


Direct: +33 6 95 14 42 13
Website: www.effibem.com

LinkedIn (en) | (fr)

**Comment by macumber** [comment]
Problem with naming the templates is 1) you might want to name the same template two different things if it is used in two different places (unless you name it PairDoubleVectorDouble or something wonderful like that) and 2) by making a class you get better named accessors, can implement logic later if needed, etc.

**Comment by macumber** [comment]
@jmarrec is this something you were working on?

**Comment by jmarrec** [comment]
> @jmarrec is this something you were working on?

I'm working on it (not done yet), will be done tomorrow.

**Comment by jmarrec** [comment]
```ruby
In [1]:
m = OpenStudio::Model::Model.new
capModFuncOfWaterFlow = OpenStudio::Model::TableMultiVariableLookup.new(m, 1)
capModFuncOfWaterFlow.setName("CapModFuncOfWaterFlow")
capModFuncOfWaterFlow.setCurveType("Quadratic")
capModFuncOfWaterFlow.setInterpolationMethod("EvaluateCurveToLimits")
capModFuncOfWaterFlow.setMinimumValueofX1(0)
capModFuncOfWaterFlow.setMaximumValueofX1(1.33)
capModFuncOfWaterFlow.setMinimumTableOutput(0.0)
capModFuncOfWaterFlow.setMaximumTableOutput(1.04)
capModFuncOfWaterFlow.setInputUnitTypeforX1("Dimensionless")
capModFuncOfWaterFlow.setOutputUnitType("Dimensionless")

capModFuncOfWaterFlow.addPoint(0.0, 0.0)
capModFuncOfWaterFlow.addPoint(0.05, 0.001)
capModFuncOfWaterFlow.addPoint(0.33333, 0.71)
capModFuncOfWaterFlow.addPoint(0.5, 0.85)
capModFuncOfWaterFlow.addPoint(0.666667, 0.92)
capModFuncOfWaterFlow.addPoint(0.833333, 0.97)
capModFuncOfWaterFlow.addPoint(1.0, 1.0)
capModFuncOfWaterFlow.addPoint(1.333333, 1.04)

In [2]: src(main)> points = capModFuncOfWaterFlow.points
Out[2]:
=> [#<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a7910
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a7898
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a7848
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a77a8
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a7708
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a7690
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a7618
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">,
 #<OpenStudio::Model::TableMultiVariableLookupPoint:0x000000044a75c8
  @__swigtype__="_p_openstudio__model__TableMultiVariableLookupPoint">]
```

**Overlaoded operator<< for `TableMultiVariableLookupPoint`:**

```ruby
In [3]: puts points[0]
Out[3]:
(x1, y) = (0, 0)
```

**New convenience function to print the data table:**

You can pass a precision too, defaults to 3 decimals

```ruby
In [4]: puts capModFuncOfWaterFlow.printTable
Out[4]:
|  n  |      x1       |       y       |
|-----+---------------+---------------|
|  1  |       0       |       0       |
|  2  |     0.05      |     0.001     |
|  3  |    0.33333    |     0.71      |
|  4  |      0.5      |     0.85      |
|  5  |    0.66667    |     0.92      |
|  6  |    0.83333    |     0.97      |
|  7  |       1       |       1       |
|  8  |    1.3333     |     1.04      |
```

**Comment by jmarrec** [comment]
@macumber can you review again please?

File: openstudiocore/src/model/TableMultiVariableLookup.cpp [hunk 1]
```diff
@@ -728,13 +778,64 @@ namespace detail {
...
-      result.push_back(p);
+      TableMultiVariableLookupPoint point(xValues, yValue);
+      result.push_back(point);
     }
 
     return result;
   }
 
+  // Helper function for printTable
+  std::string centered( std::string const& original, int targetSize ) {
```
Comment (macumber) [inline openstudiocore/src/model/TableMultiVariableLookup.cpp]: Fancy!

**Comment by macumber** [comment]
This looks awesome, nice job!

**Pull request merged**